submariner-io/submariner-operator
# build: update operator-sdk to v1.0

### New installed tools:
- controller-gen generates RBAC manifests and crds
- kustomize manages declarative configuration for each content using kustomization.yaml

### Kubebuilder-style project:
- move the CRDs from deploy/crds/ to config/crd/bases
- move the CRs from deploy/crds/ to config/samples
- move the Controller deploy/operator.yaml to config/manager/manager.yaml
- move the RBAC manifests from deploy to config/rbac/
- move pkg/apis to apis
- move pkg/controller to controllers
- move build/Dockerfile to the project's root directory
- move cmd/manager/main.go to the project's root directory
- preparations for kustomize-based deployment for prometheus, webhooks and cert-manager
- scorecard configuration to test and validate the bundle
- replace the `+k8s:deepcopy-gen:interfaces=` marker with `+kubebuilder:object:root=true` for each api type
- add the marker `+operator-sdk:csv:customresourcedefinitions` for each api
- rename each Reconciler from `Reconcile<kind>` to `<kind>Reconciler`
- replace the `Add` function with `SetupWithManager`
- operator-lib module was added in order to keep the SDK leader.Become
- save the bundle and packagemanifests for the current version
- no need for zz_generated files anymore!
- OPERATOR_NAME and POD_NAME environment variable are no longer used
- envtest framework is not yet implemented

### Add More targets to Makefile:
- `make generate` handles deepcopy and adds the license text
- `make manifests` uses controller-gen and replaces the old commands `operator-sdk generate k8s`, `operator-sdk generate openapi` and `operator-sdk generate crds`

### Dapper:
- install controller-gen and kustomize

### Other changes:
- rename the globalnet ServiceAccount to follow [RFC1123](https://tools.ietf.org/html/rfc1123)
- update the CRs
- add more entries to yamllint
- handle the embedded yamls and support the new directory structure when generating crds

### TODO:

- [ ] main.go zap.FlagSet() is no longer supported
- [ ] main.go in opeartor-sdk v1 some utilities were moved to internal
- [x] automation for adding the fields displayName and description to customresourcedefinitions.owned[]

### How to use:

#### build the operator docker image
- make operator-image

#### bundle format
- make bundle VERSION=0.6.1 CHANNELS=alpha DEFAULT_CHANNEL=alpha
- make bundle-build

#### packagemanifests format
- make packagemanifests VERSION=0.6.1 FROM_VERSION=0.5.0 CHANNEL=alpha

### Refs:
- [migration guide](https://sdk.operatorframework.io/docs/building-operators/golang/project_migration_guide)
- [controller-gen](https://book.kubebuilder.io/reference/controller-gen.html)
- [kustomize](https://kubectl.docs.kubernetes.io/pages/reference/kustomize.html)

**BREAKING CHANGE**: migrating the legacy project to the new layout. closes #678

Signed-off-by: Steve Mattar <smattar@redhat.com>
